comm/src/main/scala/coop/rchain/comm/transport/GrpcTransportClient.scala

```diff
@@ -106,27 +110,25 @@ class GrpcTransportClient(
               }
             }
       (cDef, newChannel) = ret
-      _ <- Applicative[Task].whenA(newChannel)(
-            log.info(s"Creating new channel to peer ${peer.toAddress}") >>
-              createChannel(peer) >>= cDef.complete
-          )
-      c <- cDef.get
-      // In case channel is terminated - remove current record and try one more time
-      r <- if (c.isTerminated)
-            log.info(
-              s"Channel to peer ${peer.toAddress} is terminated, removing from connections map"
-            ) >>
+      _                  <- Applicative[Task].whenA(newChannel)(createChannel(peer) >>= cDef.complete)
+      c                  <- cDef.get
+      // In case underlying gRPC transport is terminated - clean resources,
+      // remove current record and try one more time
+      r <- if (c.grpcTransport.isTerminated)
+            Task.delay(c.buferSubscriber.cancel()) >>
+              log.info(
+                s"Channel to peer ${peer.toAddress} is terminated, removing from connections map"
+              ) >>
```

Comment:
Kudos for adding canceling. I just have some concerns that can be theoretically called multiple times if multiple threads enter `if` block in the same time. Is the cancel safe or it will throw an error in that scenario.

Reply:
Cancellable sources says that `Guaranteed idempotency - calling it multiple times should have the same side-effect as calling it only once. Implementations of this method should also be thread-safe`.

The implementation of message buffer is standard Monix `PublishToOneSubject`, so it should be thread safe.
https://github.com/rchain/rchain/blob/ce06e7378531a601360d7fba52e281a61fdefe34/comm/src/main/scala/coop/rchain/comm/transport/buffer/LimitedBufferObservable.scala#L10-L20

Reply:
Nice. Thanks!